cpp/rss_aggregator.cc

```diff
@@ -367,10 +327,71 @@ int Main(int argc, char *argv[]) {
         do {
             const unsigned actual_time_slept(::sleep(static_cast<unsigned>(sleep_interval - total_time_slept)));
             CheckForSigTermAndExitIfSeen();
-            CheckForSigHupAndReloadIniFileIfSeen(&ini_file);
+            CheckForSigHupAndReloadIniFileIfSeen(ini_file);
 
             total_time_slept += actual_time_slept;
         } while (total_time_slept < sleep_interval);
         ticks += UPDATE_INTERVAL;
     }
 }
+
+
+} // unnamed namespace
+
+
+int Main(int argc, char *argv[]) {
+    if (argc < 3)
+        Usage();
+
+    bool one_shot(false);
+    if (std::strcmp(argv[1], "--one-shot") == 0) {
+        one_shot = true;
+        --argc, ++argv;
+    }
+    if (argc < 3)
+        Usage();
+
+    std::string config_file_path(UBTools::GetTuelibPath() + FileUtil::GetBasename(::progname) + ".conf");
+    if (StringUtil::StartsWith(argv[1], "--config-file=")) {
+        config_file_path = argv[1] + __builtin_strlen("--config-file=");
+        --argc, ++argv;
+    }
+
+    if (argc != 3)
+        Usage();
+
+    const std::string email_address(argv[1]);
+
+    IniFile ini_file;
+    DbConnection db_connection(ini_file);
+
+    if (not one_shot) {
+        SignalUtil::InstallHandler(SIGTERM, SigTermHandler);
+        SignalUtil::InstallHandler(SIGHUP, SigHupHandler);
+
+        if (::daemon(0, 1 /* do not close file descriptors and redirect to /dev/null */) != 0)
+            LOG_ERROR("we failed to deamonize our process!");
+    }
+
+    const std::string xml_output_filename(argv[1]);
+
+    Downloader::Params params;
+    const std::string PROXY(ini_file.getString("", "proxy", ""));
+    if (not PROXY.empty()) {
+        LOG_INFO("using proxy: " + PROXY);
+        params.proxy_host_and_port_ = PROXY;
+    }
+    Downloader downloader(params);
+
+    try {
+        return ProcessFeeds(one_shot, xml_output_filename, &ini_file, &db_connection, &downloader);
+    } catch (const std::exception &x) {
```

Comment:
Do we really want to catch std::exception or would std::runtime_error be better?